tests/test_nhg.py

```diff
@@ -384,20 +799,891 @@ def asic_route_nhg_fvs(k):
             route_ipprefix = gen_ipprefix(route_count)
             ps._del(route_ipprefix)
             route_count += 1
+        asic_routes_count -= MAX_ECMP_COUNT
 
         # Wait and check the second batch points to next hop group
         # Check ASIC DB on the count of nexthop groups used, and it should not increase or decrease
-        asic_db.wait_for_n_keys("ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP", 10)
-        keys = asic_db.get_keys("ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY")
-        k = asic_route_exists(keys, base_ipprefix)
+        asic_db.wait_for_n_keys(self.ASIC_NHG_STR, 10)
+        k = self.get_route_id(base_ipprefix, dvs)
         assert k is not None
         fvs = asic_route_nhg_fvs(k)
         assert fvs is not None
-        k = asic_route_exists(keys, last_ipprefix)
+        k = self.get_route_id(last_ipprefix, dvs)
         assert k is not None
         fvs = asic_route_nhg_fvs(k)
         assert fvs is not None
 
+        # Cleanup
+
+        # Remove second batch of routes
+        for i in range(10):
+            route_ipprefix = gen_ipprefix(MAX_ECMP_COUNT + i)
+            ps._del(route_ipprefix)
+
+        asic_db.wait_for_n_keys(self.ASIC_NHG_STR, 0)
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count)
+
+    def test_nhgorch_nhg_exhaust(self, dvs, testlog):
```

Comment:
I strongly suggest to break each test into a separate function. You can create a test class, add all the common functions in the class. It's easy to lose the track what's being tested. Moving tests into separate APIs makes it easier to maintain, add any fixes and enhance test cases.

Reply:
Could you point me to some UT that has such a test class to see how it should look like?

Reply:
Done.